Add code coverage support

eliransin opened the pull request:
This mini-set includes code coverage support for ScyllaDB, it provides:
1. Support for building ScyllaDB with coverage support.
2. Utilities for processing coverage profiling data
3. test.py support for generation and processing of coverage profiling into an lcov trace files which can later be used to produce HTML or textual coverage reports.

Refs #16323

eliransin commented:
Testing: using custom byo jenkins job: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest
An example of a report: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/41/artifact/coverage/html_report/index.html

eliransin commented:
cli refference (in case someone is interested):
```
tools/toolchain/dbuild -- test/pylib/coverage_utils.py help

usage: coverage_utils.py [-h]
                         [--log-level {CRITICAL,FATAL,ERROR,WARN,WARNING,INFO,DEBUG,NOTSET}]
                         [--concurrency CONCURRENCY]
                         {llvm-profiles,lcov-tools,help} ...

A collection of tools to handle llvm coverage data

positional arguments:
  {llvm-profiles,lcov-tools,help}
    llvm-profiles       Handle and manipulate llvm profiles
    lcov-tools          Handle and manipulate lcov tracefiles
    help                Print a full help message

options:
  -h, --help            show this help message and exit
  --log-level {CRITICAL,FATAL,ERROR,WARN,WARNING,INFO,DEBUG,NOTSET}
                        logging level of the tool (default: INFO)
  --concurrency CONCURRENCY
                        The concurrency to use for parallel operations.
                        (default: 24)
	usage: coverage_utils.py llvm-profiles [-h] {list-build-ids,merge,to-lcov} ...
	
	positional arguments:
	  {list-build-ids,merge,to-lcov}
	    list-build-ids      list build ids for the given paths
	    merge               merge and convert raw and indexed profiles to a
	                        unified profile
	    to-lcov             convert indexed profiles into lcov trace files
	
	options:
	  -h, --help            show this help message and exit
		usage: coverage_utils.py llvm-profiles list-build-ids [-h] PATH [PATH ...]
		
		positional arguments:
		  PATH        A folder to be searched for (recursively) or a specific file to
		              get an build id for
		
		options:
		  -h, --help  show this help message and exit
		usage: coverage_utils.py llvm-profiles merge [-h] [-c]
		                                             PROFILE [PROFILE ...] PROFDATA
		
		positional arguments:
		  PROFILE               a raw or indexed profile to merge into the output
		                        merged profile
		  PROFDATA              The path to which the merged indexed data will saved
		
		options:
		  -h, --help            show this help message and exit
		  -c, --clear-on-success
		                        Wether to clear the raw profiles on success (default:
		                        False)
		usage: coverage_utils.py llvm-profiles to-lcov [-h] [-c] [--exclude EXCLUDES]
		                                               [--excludes-file EXCLUDES_FILE]
		                                               -b BINARY_SEARCH_PATH
		                                               [--compilation-dir COMPILATION_DIR]
		                                               PROFILE [PROFILE ...]
		
		positional arguments:
		  PROFILE               an indexed profile to be converted to an lcov trace
		                        file
		
		options:
		  -h, --help            show this help message and exit
		  -c, --clear-on-success
		                        Wether to clear the indexed profiles on success
		                        (default: False)
		  --exclude EXCLUDES, -e EXCLUDES
		                        Regex patterns for excluding files from coverage
		                        (default: None)
		  --excludes-file EXCLUDES_FILE, --ef EXCLUDES_FILE
		                        A file containing a list of regexes to exclude
		                        (default: None)
		  -b BINARY_SEARCH_PATH, --binary-search-path BINARY_SEARCH_PATH
		                        The path or paths to search for binaries to use in the
		                        conversion. (default: None)
		  --compilation-dir COMPILATION_DIR
		                        The compilation directory for the binaries (for files
		                        compiled with relative paths mapping) (default: .)
	usage: coverage_utils.py lcov-tools [-h]
	                                    {union,diff,intersection,symmetric-dff,git-patch-coverage,genhtml,html-fixup}
	                                    ...
	
	positional arguments:
	  {union,diff,intersection,symmetric-dff,git-patch-coverage,genhtml,html-fixup}
	    union               Merges several (or single) lcov file into another
	                        trace file. If testname is given, the resulting lcov
	                        file will be tagged with this name, else if will just
	                        merge the files similarly to 'lcov -a...' command.
	                        Files can also be filtered (see 'man lcovrc')
	    diff                computes the diff between two or more coverage files
	                        (lines that are covered by first but not others)
	    intersection        computes the intersection between two or more coverage
	                        files (lines that are covered by all trace files)
	    symmetric-dff       computes the symmetric difference between two traces
	                        (line covered by either trace but not both)
	    git-patch-coverage  Transform a a source coverage tracefile into a patch
	                        coverage tracefile
	    genhtml             Generate and fixup html pages
	    html-fixup          Fix genhtml broken links for non standard file name
	                        (for example that contains '#')
	
	options:
	  -h, --help            show this help message and exit
		usage: coverage_utils.py lcov-tools union [-h] [-c] [--testname TESTNAME]
		                                          [--files-per-chunk FILES_PER_CHUNK]
		                                          [--filter]
		                                          [--exclude-line EXCLUDE_LINE]
		                                          [--exclude-start EXCLUDE_START]
		                                          [--exclude-end EXCLUDE_END]
		                                          [--exclude-branch EXCLUDE_BRANCH]
		                                          [--exclude-branch-start EXCLUDE_BRANCH_START]
		                                          [--exclude-branch-end EXCLUDE_BRANCH_END]
		                                          TRACE_FILE [TRACE_FILE ...]
		                                          OUTPUT_TRACE_FILE
		
		positional arguments:
		  TRACE_FILE            an lcov trace file to be merged into the output trace
		                        file
		  OUTPUT_TRACE_FILE     The path to which the merged lcov trace data will
		                        saved
		
		options:
		  -h, --help            show this help message and exit
		  -c, --clear-on-success
		                        Whether to clear the original lcov files upon success
		                        (default: False)
		  --testname TESTNAME   An optional testname to tag all records in the output
		                        (default: None)
		  --files-per-chunk FILES_PER_CHUNK
		                        The maximal number of files to merge at once (for
		                        performance tweaking) (default: 4)
		  --filter, -f          Apply filter to the result (see: 'man lcovrc')
		                        (default: False)
		  --exclude-line EXCLUDE_LINE, --el EXCLUDE_LINE
		                        Tag for line exclusion , empty string for None (when
		                        --filter is given) (default: LCOV_EXCL_LINE)
		  --exclude-start EXCLUDE_START, --es EXCLUDE_START
		                        Tag for line exclusion block start, empty string for
		                        None, (when --filter is given) (default:
		                        LCOV_EXCL_START)
		  --exclude-end EXCLUDE_END, --ee EXCLUDE_END
		                        Tag for line exclusion block end, empty string for
		                        None (when --filter is given) (default:
		                        LCOV_EXCL_STOP)
		  --exclude-branch EXCLUDE_BRANCH, --eb EXCLUDE_BRANCH
		                        Tag branch exclusion, empty string for None (when
		                        --filter is given) (default: LCOV_EXCL_BR_LINE)
		  --exclude-branch-start EXCLUDE_BRANCH_START, --ebs EXCLUDE_BRANCH_START
		                        Tag for branch exclusion block start, empty string for
		                        None (when --filter is given) (default:
		                        LCOV_EXCL_BR_START)
		  --exclude-branch-end EXCLUDE_BRANCH_END, --ebe EXCLUDE_BRANCH_END
		                        Tag for branch exclusion block end, empty string for
		                        None (when --filter is given) (default:
		                        LCOV_EXCL_BR_STOP)
		usage: coverage_utils.py lcov-tools diff [-h] --output-trace OUTPUT_TRACE
		                                         [--files-per-chunk FILES_PER_CHUNK]
		                                         base_tracefile diff_tracefiles
		                                         [diff_tracefiles ...]
		
		positional arguments:
		  base_tracefile        The base line trace - the file which we want to diff
		                        with all others
		  diff_tracefiles       The tracefiles to subtracted from the base trace
		
		options:
		  -h, --help            show this help message and exit
		  --output-trace OUTPUT_TRACE, -o OUTPUT_TRACE
		                        The output file to write the result into (default:
		                        None)
		  --files-per-chunk FILES_PER_CHUNK
		                        The max number of files to merge at once (for
		                        performance tweaking) (default: 4)
		usage: coverage_utils.py lcov-tools intersection [-h] --output-trace
		                                                 OUTPUT_TRACE
		                                                 [--files-per-chunk FILES_PER_CHUNK]
		                                                 tracefiles [tracefiles ...]
		
		positional arguments:
		  tracefiles            The tracefiles to subtract from the base trace
		
		options:
		  -h, --help            show this help message and exit
		  --output-trace OUTPUT_TRACE, -o OUTPUT_TRACE
		                        The output file to write the result into (default:
		                        None)
		  --files-per-chunk FILES_PER_CHUNK
		                        The max number of files to merge at once (for
		                        performance tweaking) (default: 4)
		usage: coverage_utils.py lcov-tools symmetric-dff [-h] --output-trace
		                                                  OUTPUT_TRACE
		                                                  tracefiles tracefiles
		
		positional arguments:
		  tracefiles            The tracefiles to be subtracted
		
		options:
		  -h, --help            show this help message and exit
		  --output-trace OUTPUT_TRACE, -o OUTPUT_TRACE
		                        The output file to write the result into (default:
		                        None)
		usage: coverage_utils.py lcov-tools git-patch-coverage [-h] --output-trace
		                                                       OUTPUT_TRACE
		                                                       [--tracefile TRACEFILE]
		                                                       [--base-commit BASE_COMMIT]
		                                                       --output-dir OUTPUT_DIR
		                                                       [--dirty]
		                                                       [--clear-output-dir]
		                                                       [--pseudo-patch-for-uncovered]
		                                                       [--merge]
		                                                       [--merge-parent {1,2}]
		
		options:
		  -h, --help            show this help message and exit
		  --output-trace OUTPUT_TRACE, -o OUTPUT_TRACE
		                        The output file to write the result into (default:
		                        None)
		  --tracefile TRACEFILE, -t TRACEFILE
		                        The tracefile to transform (default: None)
		  --base-commit BASE_COMMIT, -b BASE_COMMIT
		                        The base commit for the patch coverage (default: None)
		  --output-dir OUTPUT_DIR, -d OUTPUT_DIR
		                        The directory to create the patches in (default: None)
		  --dirty               Wether to include a final meta patch which is the
		                        uncommitted changes to the environment,if this
		                        parameter is not given and there are uncommitted
		                        changes in the repo, the command will fail. (not
		                        including untracked files) (default: False)
		  --clear-output-dir, -c
		                        Remove any previous .patch file in output dir
		                        (default: False)
		  --pseudo-patch-for-uncovered, -p
		                        Create a pseudo patch for uncovered patch files and
		                        make them appear as 100% covered (default: False)
		  --merge, -m           Coverage report for merge commit (default: False)
		  --merge-parent {1,2}, --mp {1,2}
		                        The parent of the merge commit to generate coverege
		                        for (default: 2)
		usage: coverage_utils.py lcov-tools genhtml [-h] --output-dir OUTPUT_DIR
		                                            [--verbose] [--title TITLE]
		                                            [--no-legend]
		                                            [--no-function-coverage]
		                                            [--no-branch-coverage]
		                                            [--no-cpp-demangle]
		                                            [--ignore-errors] [--negate]
		                                            [--high-limit HIGH_LIMIT]
		                                            [--med-limit MED_LIMIT]
		                                            tracefile [tracefile ...]
		
		positional arguments:
		  tracefile             The tracefiles to generate the reports from
		
		options:
		  -h, --help            show this help message and exit
		  --output-dir OUTPUT_DIR, -o OUTPUT_DIR
		                        The folder in which to generate the html report
		                        (default: None)
		  --verbose, -v         Output everything instead of just warnings and errors
		                        (default: False)
		  --title TITLE, -t TITLE
		                        The title of the coverage run for example: 'Unit tests
		                        run' (default: None)
		  --no-legend, --nl     Don't create legend (default: False)
		  --no-function-coverage, --nf
		                        Don't report function coverage (default: False)
		  --no-branch-coverage, --nb
		                        Don't report branch coverage (default: False)
		  --no-cpp-demangle, --nd
		                        Don't demangle function names (default: False)
		  --ignore-errors, -i   Ignore 'source wasn't found' errors (default: False)
		  --negate, -n          Make the report missed coverage centric instead of
		                        coverage centric (default: False)
		  --high-limit HIGH_LIMIT, --hl HIGH_LIMIT
		                        The high limit for reporting (high coverage) (default:
		                        90)
		  --med-limit MED_LIMIT, --ml MED_LIMIT
		                        The medium limit for reporting (medium coverage)
		                        (default: 75)
		usage: coverage_utils.py lcov-tools html-fixup [-h] --html-dir HTML_DIR
		
		options:
		  -h, --help            show this help message and exit
		  --html-dir HTML_DIR, -d HTML_DIR
		                        The folder containing the html report generated by
		                        htmlgen (default: None)
	usage: coverage_utils.py help [-h]
	
	options:
	  -h, --help  show this help message and exit

```

eliransin commented:
@michoecho I have put you as a reviewer since you dealt with profiles in the past so maybe this is of interest to you.
@denesb as the author of `coverage` build mode, I need your blessing :slightly_smiling_face:

eliransin commented:
for completeness here is a patch coverage report for the top 15 commits from the run in https://github.com/scylladb/scylladb/pull/16784#issuecomment-1892290025 :

[patch_coverage.tar.gz](https://github.com/scylladb/scylladb/files/13940202/patch_coverage.tar.gz)

The way to generate this is:
1. cd into the source dir
2. `mkdir coverage`
3. Download `unified.info` from the jenkins job into `coverage`
4. `test/pylib/coverage_utils.py lcov-tools git-patch-coverage -o coverage/patch_coverage.info -t coverage/unified.info -b HEAD~15 -d patches -p` - this generates patch coverage info in `coverage/patch_coverage.info`
5. `test/pylib/coverage_utils.py lcov-tools genhtml -o coverage/patch_coverage -t "patch coverage for HEAD~15" -v coverage/patch_coverage.info` - this generates an html report in `coverage/patch_coverage` dir.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5790/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7159/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5790/)
#### Build Details:
- Duration: 2 hr 11 min
- Builder: i-05363d5bfddd95145 (m5d.12xlarge)

denesb commented on test.py:
```diff
@@ -145,7 +145,15 @@ def __init__(self, path: str, cfg: dict, options: argparse.Namespace, mode: str)
+            # this way is that the storage will not be bloated with coverage files (each can weigh 10s of MBs so for several
+            # thousands of tests it can easily reach 10 of GBs)
+            # ref: https://clang.llvm.org/docs/SourceBasedCodeCoverage.html#running-the-instrumented-program
+            self.base_env["LLVM_PROFILE_FILE"] = os.path.join(options.tmpdir,self.mode,"coverage",self.name,"%m.profraw")
```
nit: space after comma

denesb commented on test.py:
```diff
@@ -274,7 +282,8 @@ async def add_test(shortname) -> None:
                 task.cancel()
             await asyncio.gather(*pending, return_exceptions=True)
             raise
-
```
nit: keep the empty line

denesb commented on test.py:
```diff
@@ -1256,7 +1265,13 @@ def parse_cmd_line() -> argparse.Namespace:
                              "By default, the marker filter is not applied and all tests will be run without exception."
                              "To exclude e.g. slow tests you can write --markers 'not slow'.")
-
+    parser.add_argument('--coverage', action='store_true', default = False,
```
no space around the `=`, when passing kwargs

denesb commented on test.py:
```diff
@@ -1256,7 +1265,13 @@ def parse_cmd_line() -> argparse.Namespace:
+                             "lcov file per suite, lcov file per mode, and an lcov file for the entire run, "
+                             "The lcov files can eventually be used for generating coverage reports")
+    parser.add_argument("--coverage-mode",action = 'append', type=str, dest = "coverage_modes",
+                        help = "Collect and process coverage only for the modes specified. implies: --coverage, defalt: All built modes")
```
no space around the `=`

denesb commented on test.py:
```diff
@@ -1294,7 +1309,19 @@ def parse_cmd_line() -> argparse.Namespace:
         except Exception:
             print(palette.fail("Failed to read output of `ninja mode_list`: please run ./configure.py first"))
             raise
-
```
keep the empty line

denesb commented on test.py:
```diff
@@ -1256,7 +1265,13 @@ def parse_cmd_line() -> argparse.Namespace:
+                             "Will route the profiles to `tmpdir`/mode/coverage/`suite` and post process them in order to generate "
+                             "lcov file per suite, lcov file per mode, and an lcov file for the entire run, "
+                             "The lcov files can eventually be used for generating coverage reports")
+    parser.add_argument("--coverage-mode",action = 'append', type=str, dest = "coverage_modes",
```
space after comma

denesb commented on test.py:
```diff
@@ -1294,7 +1309,19 @@ def parse_cmd_line() -> argparse.Namespace:
+            args.coverage = False
+    elif args.coverage_modes:
+        if "coverage" in args.coverage_modes:
+            raise RuntimeError("'coverag' mode is not allowed in --coverage-mode")
```
coverage

denesb commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1286 @@
+import sys
+import inspect
+#So the module can be imported outside of this directory
+sys.path.insert(0,os.path.join(os.path.dirname(__file__)))
```
I don't understand how this enables importing this file?

eliransin replied:
`coverage_utils` imports lcov_utils which resided in the same directory.
However, for example if you `import coverage_utils` in `test.py` without this "trick",
the import will fail because python will claim that it couldn't find lcov_utils module.
python default behavior for searching modules is:
1. `cwd` (which is the main source tree when running test.py)
2. The system defined directories

according to these rules, python will not search in `test/pylib` which is where lcov_utils resides. Adding this line tells python to also search in the local directory where `coverage_utils.py` resides. And then removes this rule immediately after finding the local module so following imports in the external file (which in this example is `test.py`) will not import by accident additional module from that directory that happen to have names identical to the system modules. (look at the 2 lines that comes right after this line)

eliransin replied:
@denesb let me know if it answers you question.

denesb reviewed (commented):
Thanks for giving the coverage some much needed love and transforming it into something that is actually useful.

Is the coverage mode still needed? If not, we should remove it (can be a follow-up).

The interaction between `test.py`'s `--coverage` and `--coverage-mode` is somewhat confusing. Maybe we only need the latter.
The proposed modus operandi sounds good to me.

I left some nits around, they are not important. I didn't review the new libraries, they are massive and I don't know the intricacies of the format. Are these imported from some external project or did you write them?

eliransin commented:
> Thanks for giving the coverage some much needed love and transforming it into something that is actually useful.
> 
> Is the coverage mode still needed? If not, we should remove it (can be a follow-up).

I believe it is not needed anymore, I intended to suggest this as a followup but wasn't sure if and how it is needed.

> 
> The interaction between `test.py`'s `--coverage` and `--coverage-mode` is somewhat confusing. Maybe we only need the latter. The proposed modus operandi sounds good to me.

I wanted to make the usage as simple as possible, so a user can choose between producing coverage for only some modes (individually stating modes --coverage-mode while running the tests on more modes) and all of the modes that the tests run on - @denesb do you have an easier to interact option in mind? The only thing that I can think of is adding to --coverage-mode a pseudo mode, "all" which will do coverage on all modes.  I tend to like the current implementation better since it saves typing but not very opinionated about it :slightly_smiling_face: 
> 
> I left some nits around, they are not important. I didn't review the new libraries, they are massive and I don't know the intricacies of the format. Are these imported from some external project or did you write them?

Will take care of it.

fruch commented:
for who want a quick look at the results:
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/41/artifact/coverage/html_report/index.html

@eliransin, few questions:
 
1) can you explain what is the `*-only` results, it's not exactly clear what one should be looking when reading those.
2) where is equivalent scylla-pkg PR that operate all of this ?
3) suggestion, maybe to run it on a full dtest run, not just gating (assuming this would be running on a lower cadence then gating tests would), and later we can compare and target which test should be added to gating (and which tests needs to be written)

eliransin commented:
@fruch 
> for who want a quick look at the results: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/41/artifact/coverage/html_report/index.html
> 
> @eliransin, few questions:
> 
> 1. can you explain what is the `*-only` results, it's not exactly clear what one should be looking when reading those.
Those coverage-diff reports:
the `dtest-only` answers the question what code is covered by dtest but not by unit-tests(test.py)
the `unit-test-only` answers the question what code is covered by unit-test but not by dtest.
> 2. where is equivalent scylla-pkg PR that operate all of this ?
It is referenced in this issue, I don't include a link since it is a private repo (you can scroll and find it)
> 3. suggestion, maybe to run it on a full dtest run, not just gating (assuming this would be running on a lower cadence then gating tests would), and later we can compare and target which test should be added to gating (and which tests needs to be written)
I actually fired up a run of full dtest yesterday but it failed on artifact publish, so I reran it, will share the results when it is over.

denesb commented:
> I believe it is not needed anymore, I intended to suggest this as a followup but wasn't sure if and how it is needed.

I agree, we can remove it in a follow-up.

> I wanted to make the usage as simple as possible, so a user can choose between producing coverage for only some modes (individually stating modes --coverage-mode while running the tests on more modes) and all of the modes that the tests run on - @denesb do you have an easier to interact option in mind? The only thing that I can think of is adding to --coverage-mode a pseudo mode, "all" which will do coverage on all modes. I tend to like the current implementation better since it saves typing but not very opinionated about it 🙂

Ok, I'm fine with it as-is. I thought the interaction is more complex, when I looked at the `test.py` code, but maybe I didn't understand it well.
 
> > I left some nits around, they are not important. I didn't review the new libraries, they are massive and I don't know the intricacies of the format. Are these imported from some external project or did you write them?
> 
> Will take care of it. I wrote the libraries myself 🙂

Nice. The example report linked by @fruch looks really cool (apart from all the red lines due to our low coverage -- we need to start working on bumping those up into the "green").

eliransin commented:
@roydahan found a possible bug in coverage diff and I am working on this so halt with the merge (if someone even planned to merge it right now :slightly_smiling_face: )

roydahan commented:
@eliransin It's a tiny issue.
I wouldn't hold the merge just for this, we can fix it in a following PR.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5803/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7168/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5803/)
#### Build Details:
- Duration: 1 hr 35 min
- Builder: spider6.cloudius-systems.com

avikivity commented:
Please add something to docs/dev to explain how to work with this.

avikivity commented on configure.py:
```diff
@@ -778,8 +778,11 @@ def find_ninja():
 arg_parser.add_argument('--date-stamp', dest='date_stamp', type=str,
                         help='Set datestamp for SCYLLA-VERSION-GEN')
 arg_parser.add_argument('--use-cmake', action='store_true', help='Use CMake as the build system')
+arg_parser.add_argument('--coverage', action='store_true', help = 'Compile scylla with coverage instrumentation')
```
We have --mode=coverage, no?

avikivity replied:
But you're adding three new coverage modes, with an extra flag.

Why not fix the coverage mode?

avikivity commented on configure.py:
```diff
@@ -1418,6 +1421,15 @@ def find_ninja():
 ]) | pure_boost_tests
 
+
+COVERAGE_INST_FLAGS = ['-fprofile-instr-generate','-fcoverage-mapping',f'-fprofile-list=./{PROFILES_LIST_FILE_NAME}']
```
space after comma

avikivity commented on configure.py:
```diff
@@ -1418,6 +1421,15 @@ def find_ninja():
+
+COVERAGE_INST_FLAGS = ['-fprofile-instr-generate','-fcoverage-mapping',f'-fprofile-list=./{PROFILES_LIST_FILE_NAME}']
+if args.coverage:
+    for _ ,mode in filter(lambda m: m[0] != "coverage" ,modes.items()):
```
space after comma ,not before

avikivity commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015-present ScyllaDB
```
I'm sure it didn't take 9 years to write.

eliransin replied:
will fix

avikivity commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+def traced_func(f):
+    """
+    A tracing helper which logs function name and parameters so it
+    is easy to debug this module.
```
I don't understand, is this to debug coverage_utils.py itself?

eliransin replied:
Yes, as explained in the doc-string, if the logging level is set to debug, this will dump the parameters given to
any function that was annotated with this.

avikivity replied:
Please remove it, it makes the file complicated.

I'm sure there are python debuggers and tracers.

avikivity commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+
+_create_subprocess_exec_sig = inspect.signature(asyncio.create_subprocess_exec)
+@wraps(asyncio.create_subprocess_exec)
+async def create_subprocess_exec(*args,logger =COVERAGE_TOOLS_LOGGER ,**kwargs):
```
Wild spacing again.

avikivity commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+    elif isinstance(concurrency,asyncio.Semaphore):
+        return concurrency
+    else:
+        return asyncio.Semaphore(1)
```
I don't understand why this is needed, doesn't python already do this?

Don't we use just one concurrency library? Can't we write asyncio.Semaphore(concurrency) at the call site instead of this function?

eliransin replied:
This function is to allow an external semaphore to be given to the library, imagine a case where you want a
function to share the concurrency with other functions in your program.
It is not mandatory but allows greater flexibility when used as a library.

avikivity commented on coverage_excludes.txt:
```diff
@@ -0,0 +1,12 @@
+# This is a simple file containing regx patterns that should be
```
It's good that the file explains it is a simple file.

eliransin replied:
Will write a less stupid comment :slightly_smiling_face:

eliransin commented on configure.py:
```diff
@@ -1418,6 +1421,15 @@ def find_ninja():
+if args.coverage:
+    for _ ,mode in filter(lambda m: m[0] != "coverage" ,modes.items()):
+        mode['cxx_ld_flags'] += ' ' + ' '.join(COVERAGE_INST_FLAGS)
+        mode['cxx_ld_flags'] = mode['cxx_ld_flags'].strip()
```
to clear a leading space in case `cxx_ld_flags` was an empty string before adding the coverage flags.
I could leave it as is but wanted to have a clean `cxx_ld_flags` string.

michoecho reviewed (commented):
That's a whole lot of python to review. I only reviewed a small fraction of it so far and only pointed out some surface-level things, didn't even taste the "meat".

But this is just an internal helper script. Maybe we shouldn't try to review it deeply? If it works, it works.

eliransin commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+    """ A wrapper around asyncio.gather that limits concurrency for the submitted tasks.
+    This function takes parameters as asyncio.gather with an additional mandatory `semaphore` param that
+    is the semaphore that will limit the tasks concurrency.
+    Note: The function looks a bit convoluted, but it is actually a nuance that arises from the semaphore.
+    Since asyncio.gather, doesn't stop already submitted tasks, it can cause deadlock on the semaphore. So
+    we should make sure that in case of a failure the all tasks are terminated unconditionally.
```
I am not sure it will work well with the `return_exceptions` option of  gather.
But I guess I could just try it out on some toy example first :slightly_smiling_face:

eliransin replied:
As I suspected, the proposed implementation just throws an exception and cancel tasks even if we pass
`return_exceptions = True` to `gather`

michoecho replied:
True, that one won't work with `return_exceptions`. But this one will:
```
async def gather_limited_concurrency(*args: Coroutine, semaphore: asyncio.Semaphore, **kwargs) -> list[Any]:
    async with asyncio.TaskGroup() as tg:
        tasks = [tg.create_task(with_semaphore(semaphore, x)) for x in args]
    return await asyncio.gather(*tasks, **kwargs)
```
You aren't using `return_exceptions` anywhere, though.

michoecho replied:
Also, task group has different semantics: it doesn't just cancel the subtasks, but also waits for them to end. But I think that's a good thing.

michoecho replied:
Anyway, yes, if we want `return_exceptions` to work, then the above doesn't work. You'd have to wrap each coroutine in a try..except.

eliransin commented:
Working on the fixes and comments, but in the meantime, as requested by @fruch and @roydahan , here is a report with the same tests that we run in detst-release :slightly_smiling_face: 
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/49/artifact/coverage/html_report/index.html

eliransin commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+    else:
+        return asyncio.Semaphore(1)
+
+async def verify_file_type(f : PathLike, file_type: Union[FileType,Iterable[FileType]]) -> bool:
+    """Verify that a file is of certain type
+
+    Args:
+        f (PathLike): The path to the file whose type needs to be verified
+        file_type (Union[FileType,Iterable[FileType]]): A collection of filetypes or a single
+        file type for the file to be verified against.
+
+    Returns:
+        bool: True if the file type matches one of `file_type` False otherwise.
+    """
+    if not isinstance(file_type,IterableType):
+        file_type = [ file_type ]
+    file_type = list(file_type)
+    return await FileType.get_file_type(f) in file_type
```
The reason this helper exists is because there is no "async lambda", Since get_file_type uses system process under the hood. The idea is to parallelize multiple "verifies".
If this function doesn't exists, lines like this will not be possible:
```
 profiles_filter = await gather_limited_concurrency(
        *(
            verify_file_type(f, [FileType.RAW_PROFILE, FileType.INDEXED_PROFILE])
            for f in profiles
        ),
        concurrency = sem,
    )
```
However, I do see a pattern of the usage of this function, I always, paralelly verify file types and then
use the other function you didn't like (:slightly_smiling_face: ), `filter_from_list` to get only the files I need.
So actually I can probably can unify them to one function with `verify_file_type` being an internal function in this.
I think this is what I'll do.

eliransin replied:
Removed all of those functions as I found out they are used only in very specific places.

eliransin commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+PathLikeCollection = Iterable[PathLike]
+ConcurrencyParam = Optional[Union[int,asyncio.Semaphore]]
+PathCollection = Iterable[Path]
+PathLikeOrPathLikeCollection = Union[PathLike,Iterable[PathLike]]
```
You have a point :) I will change this.

denesb commented:
> But this is just an internal helper script. Maybe we shouldn't try to review it deeply? If it works, it works.

Yes, we usually have much more relaxed criteria for internal helper scripts.

eliransin commented:
Pushed a new version with:
1. Python code format and style fixes (will resolve review comments accordingly)
2. Most (if not all) comments addressed

Will give it another byo run on jenkins to see everything is still functional and report.

eliransin commented:
> Please add something to docs/dev to explain how to work with this.

New version also contains a documentation commit.

eliransin commented on test/pylib/coverage_utils.py:
```diff
@@ -0,0 +1,1293 @@
+    """
+    if invert:
+        filter = [not filt for filt in filter]
+    return [x for x,y in zip(objects,filter) if y]
```
Removed

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5840/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7207/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5840/)
#### Build Details:
- Duration: 2 hr 23 min
- Builder: i-0ea0320ce7dd6a017 (m5d.12xlarge)

avikivity commented:
> That's a whole lot of python to review. I only reviewed a small fraction of it so far and only pointed out some surface-level things, didn't even taste the "meat".
> 
> But this is just an internal helper script. Maybe we shouldn't try to review it deeply? If it works, it works.

I do less-deep reviews of helpers, but the downside is it is harder to maintain them later.

Also, shouldn't this be integrated in test.py?

eliransin commented:
> > That's a whole lot of python to review. I only reviewed a small fraction of it so far and only pointed out some surface-level things, didn't even taste the "meat".
> > But this is just an internal helper script. Maybe we shouldn't try to review it deeply? If it works, it works.
> 
> I do less-deep reviews of helpers, but the downside is it is harder to maintain them later.
> 
> Also, shouldn't this be integrated in test.py?

@avikivity It is integrated into test.py as part of this PR, the docs I added explains how to use it and generate reports from it.

eliransin commented:
Pushed another version:
1. rebased on latest master branch
2. Fixed wild spacing and coding style issues all over
3. Removed the concurrency->semaphore function and instead the library will always take a semaphore or will default
to no concurrency at all.
4. Added options to keep intermediate results in test.py so in the future users can choose if they want to role their own processing (using the llvm tools) instead of our processing pipeline.

Manually tested test.py workflow.
Will now run a final test job to make sure nothing is broken for main use case (currently byo).

eliransin commented:
Pushed a fix for html fixup (important for patch coverage) tested manually

eliransin commented:
The byo job passe : https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/53/
Report is in: 
https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/eliran/job/byo/job/byo_build_tests_dtest/53/artifact/coverage/html_report/index.html

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5876/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7228/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/5876/)
#### Build Details:
- Duration: 2 hr 22 min
- Builder: i-078fd8c2054a2d930 (m5d.12xlarge)

eliransin commented:
@avikivity @denesb @michoecho all comments addressed or answered, is there anything else I should do in order to
get this merged?

denesb commented:
From my end, this is good to go in, but I'll leave some time for the others to answer too.

michoecho commented:
I didn't review most of it due to the length, but I think that a full review or further nitpicking would be counterproductive in this case, so I will refrain from any other comments. If it works as expected, looks good to me.

denesb commented:
Nobody objected, so the PR is queued.

avikivity commented:
> Thanks for giving the coverage some much needed love and transforming it into something that is actually useful.
> 
> Is the coverage mode still needed? If not, we should remove it (can be a follow-up).


I think this should have been folded into coverage mode, not add yet another orthogonal switch.


As a side effect, cmake is now broken wrt coverage.

denesb commented:
> I think this should have been folded into coverage mode, not add yet another orthogonal switch.

Coverage mode will be removed soon. I think a switch is better then a separate build mode. The separate build mode was one of the reasons nobody used coverage.
